File: atlas/examples/kotlin/domain/src/main/kotlin/dev/atlas/example/domain/catalog/DefaultCatalogTitlePolicy.kt
package dev.atlas.example.domain.catalog

/**
 * Applies the catalog's default whitespace and leading-capital title policy.
 */
class DefaultCatalogTitlePolicy : CatalogTitlePolicy {
    /**
     * Normalizes one candidate while rejecting blank titles.
     *
     * @param candidate User-owned title text.
     * @return Collapsed, leading-capital title text.
     */
    override fun normalize(candidate: String): String {
        val collapsed = candidate.trim().replace(Regex("\\s+"), " ")
        require(collapsed.isNotEmpty()) { "Catalog item title must not be blank." }
        return collapsed.replaceFirstChar { character -> character.titlecase() }
    }
}

File: atlas/examples/typescript/packages/domain/src/catalog/DefaultCatalogTitlePolicy.ts
import type { CatalogTitlePolicy } from './CatalogTitlePolicy.js';

/**
 * Applies the catalog's default whitespace and leading-capital title policy.
 */
export class DefaultCatalogTitlePolicy implements CatalogTitlePolicy {
  /**
   * Normalizes one candidate while rejecting blank titles.
   *
   * @param candidate - User-owned title text.
   * @returns Collapsed, leading-capital title text.
   */
  public normalize(candidate: string): string {
    const collapsed = candidate.trim().replaceAll(/\s+/g, ' ');
    if (collapsed.length === 0) {
      throw new Error('Catalog item title must not be blank.');
    }
    return `${collapsed[0]?.toLocaleUpperCase()}${collapsed.slice(1)}`;
  }
}
